### PR TITLE
Long variable subtract suffix

### PR DESCRIPTION
**Pull request by frankdekker**
Type: feature
Breaking change: no

At our company we have the LongVariable rule set to the default 20 characters. In certain scenario's the suffix of the variable gives just a little bit more information. Usually we resort to abbrevations as overall we don't want to increase the maximum for the LongVariable Rule.

I've added a "subtract-suffixes" property to the LongVariable Rule which will subtract the first matching suffix of the comma-separated list from the variable name before calculating the length.

**Example:** 
'subtract-suffixes'='Repository,Factory,Provider'

If the variable name is on the left, the variable name will be counted as the right:
- $permissionsRepository   ->  $permissions
- $permissionsRepositoryFactory -> $permissionsRepository
- $permissionsServicePool -> $permissionsServicePool

At most 1 (one) suffix will be subtracted.

- Added test to cover my changes
- Updated the documentation to reflect the 'subtract-suffixes' property

I've one question about the tests. All tests for LongVariableTest had a @ since annotation, but as I don't know in which release this will go, what has to be added?

### CODE REVIEW

**Comment by tvbeek** [comment]
@frankdekker Thanks for the PR, very nice to see that you not only contributed a change but as far as I see in a first look it is complete with tests and documentation. Very good!

I need some more time to review it on another moment, but my first impression is good.

About your question, I currently suspect that this will be added in 2.9.0 and place that as the milestone.

File: src/main/php/PHPMD/Rule/Naming/LongVariable.php [hunk 1]
```diff
@@ -169,4 +196,24 @@ protected function isNotProcessed(AbstractNode $node)
...
+     *
+     * @return string[]
+     */
+    private function getSubtractSuffixList()
+    {
+        if ($this->subtractSuffixes !== null) {
+            return $this->subtractSuffixes;
+        }
+
+        try {
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Naming/LongVariable.php:211]: Not part of this PR but I think we should introduce a default/fallback value for ``getStringProperty()``. This way we can get rid of all these "try, getStringProperty(), catch & use fallback value"" constructs.

Comment (frankdekker) [reply src/main/php/PHPMD/Rule/Naming/LongVariable.php:211]: I agree, I was first looking for a hasProperty function, and after i couldn't find one, checked the other rules how it was solved there. As this was my first steps with the project i didn't feel comfortable to change the core :)

Comment (ravage84) [reply src/main/php/PHPMD/Rule/Naming/LongVariable.php:211]: @frankdekker see my new PR https://github.com/phpmd/phpmd/pull/750

File: src/main/php/PHPMD/Rule/Naming/LongVariable.php [hunk 2]
```diff
@@ -169,4 +196,24 @@ protected function isNotProcessed(AbstractNode $node)
...
+            $suffixes = $this->getStringProperty('subtract-suffixes');
+        } catch (\OutOfBoundsException $e) {
+            return $this->subtractSuffixes = array();
+        }
+
+        return $this->subtractSuffixes = array_map('trim', explode(',', $suffixes));
```
Comment (tvbeek) [inline src/main/php/PHPMD/Rule/Naming/LongVariable.php:217]: Maybe we should not only trim the values but also filter it so we don't need to check if the length > 0 (On line 139 https://github.com/phpmd/phpmd/pull/747/files#diff-c8f61daf277f66922e2b67f8d09af86cR139 )

Or do I miss a usecase?

Comment (frankdekker) [reply src/main/php/PHPMD/Rule/Naming/LongVariable.php]: Agree. No need for empty suffixes. Updated

File: src/main/php/PHPMD/Rule/Naming/LongVariable.php [hunk 3]
```diff
@@ -214,6 +214,14 @@ private function getSubtractSuffixList()
             return $this->subtractSuffixes = array();
         }
 
-        return $this->subtractSuffixes = array_map('trim', explode(',', $suffixes));
+        return $this->subtractSuffixes = array_filter(
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Naming/LongVariable.php:217]: https://github.com/phpmd/phpmd/pull/747/files/7749aeb91e808837c8de291cb46643e6a54bc9ec..0c11a52fd8f90bda9db1613172881cb5caee816f#diff-c8f61daf277f66922e2b67f8d09af86cR211-R227

This try-getStringProperty-catchempty-array-array-map-with-trimmed-explode-and-filter could be extracted to a new ``AbstractRule::getListProperty()``

File: src/main/php/PHPMD/Rule/Naming/LongVariable.php [hunk 4]
```diff
@@ -136,7 +136,7 @@ private function getStringLength($variableName, array $subtractSuffixes)
     {
         foreach ($subtractSuffixes as $suffix) {
             $suffixLength = strlen($suffix);
-            if ($suffixLength > 0 && substr($variableName, -$suffixLength) === $suffix) {
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Naming/LongVariable.php]: Well, before the code was slightly more defensive. Since the ``getStringLength()`` method doesn't retrieve the list, it theoretically still could contain empty suffixes. E.g. in a subclass.

**Comment by ravage84** [comment]
@frankdekker thank you for your work.

Would you be interested to implement a new ``LongClassName`` rule which also supports the suffix property of this PR?

**Comment by frankdekker** [comment]
> @frankdekker thank you for your work.
> 
> Would you be interested to implement a new `LongClassName` rule which also supports the suffix property of this PR?

@ravage84 
That would be interesting rule. Sure. You have a recommendation of the default "max" length for a class name?

**Comment by ravage84** [comment]
> That would be interesting rule. Sure. You have a recommendation of the default "max" length for a class name?

@frankdekker not really.

I didn't find any such rule in [Java's PMD](https://pmd.github.io/latest/).

In PhpStorm the default settings seem to be these:

grafik

**Comment by frankdekker** [comment]
@ravage84 
Interesting example from phpstorm. Instead of making it a LongClassName rule, could do a ClassNamingConventions rule. Similar to ConstantNamingConventions. 

**With properties:**
- pattern:  (default: [A-Z][A-Za-z0-9]*)
- min:   (default: 4)
- max:   (default: 20)
- subtract-suffixes: (default: none)

Thoughts?

**Comment by ravage84** [comment]
@frankdekker moved discussion to https://github.com/phpmd/phpmd/issues/763. Please answer there.
